pkg/chaosdaemon/tasks/task_manager.go

```diff
@@ -0,0 +1,278 @@
+//	Sum task configs on same PID into one.
+// 	Create task.
+// 	Assign or update task.
+// SO if developers wants to use functions in TaskManager ,
+// their imported TaskConfig need to implement interface TaskExecutor.
+// If developers wants to recover task successfully,
+// the task must implement Recoverable.
+// If not implement ,
+// the Recover function will return a ErrNotImplement("Recoverable") error.
+// IMPORTANT: We assume task config obey that TaskConfig.data A,B. A.Add(B)
+// is approximately equal to B.Add(A)
+type TaskManager struct {
+	taskConfigManager TaskConfigManager
+	taskMap           map[PID]Injectable
+
+	logger logr.Logger
+}
+
+func NewTaskManager(logger logr.Logger) TaskManager {
+	return TaskManager{
+		NewTaskConfigManager(),
+		make(map[PID]Injectable),
+		logger,
+	}
+}
+
+func (cm TaskManager) CopyTaskConfigManager() TaskConfigManager {
+	tm := NewTaskConfigManager()
+	for uid, task := range cm.taskConfigManager.TaskConfigMap {
+		tm.TaskConfigMap[uid] = task
+	}
+	return tm
+}
+
+func (cm TaskManager) CopyTaskMap() map[PID]Injectable {
+	pm := make(map[PID]Injectable)
+	for pid, chaosOnProcess := range cm.taskMap {
+		cm.taskMap[pid] = chaosOnProcess
+	}
+	return pm
+}
+
+func (cm TaskManager) GetConfigWithUID(id UID) (interface{}, error) {
+	return cm.taskConfigManager.GetConfigWithUID(id)
+}
+
+func (cm TaskManager) GetTaskWithPID(pid PID) (Injectable, error) {
+	p, ok := cm.taskMap[pid]
+	if !ok {
+		return nil, chaoserr.NotFound("PID")
+	}
+	return p, nil
+}
+
+func (cm TaskManager) GetUIDsWithPID(pid PID) []UID {
+	return cm.taskConfigManager.GetUIDsWithPID(pid)
+}
+
+// Update the task with a same UID, PID and new task config.
+// If it comes a UID not injected , Update will return ChaosErr.NotFound("UID").
+// If it comes the import PID of task do not equal to the last one,
+// Update will return ErrUpdateTaskConfigWithPIDChanges.
+func (cm TaskManager) Update(uid UID, pid PID, config TaskExecutor) error {
+	oldTask, err := cm.taskConfigManager.UpdateTask(uid, NewTaskConfig(pid, config))
+	if err != nil {
+		return err
+	}
+	err = cm.commit(uid, pid)
+	if err != nil {
+		_, _ = cm.taskConfigManager.UpdateTask(uid, oldTask)
+		return err
+	}
+	return nil
+}
+
+// Create the first task,
+// the New function of TaskExecutor:Creator will only be used here.
+// values is only the import parameter of New function in TaskExecutor:Creator.
+// If it comes a task are already be injected on the PID,
+// Create will return ChaosErr.ErrDuplicateEntity.
+func (cm TaskManager) Create(uid UID, pid PID, config TaskExecutor, values interface{}) error {
+	if _, ok := cm.taskMap[pid]; ok {
+		return errors.Wrapf(chaoserr.ErrDuplicateEntity, "create")
+	}
+
+	err := cm.taskConfigManager.AddTask(uid, NewTaskConfig(pid, config))
+	if err != nil {
+		return err
+	}
+
+	processTask, err := config.New(values)
+	if err != nil {
+		_ = cm.taskConfigManager.RecoverTask(uid)
```

Comment:
Why will we recover the task here? It seems that it hasn't been injected.

Reply:
```
	err := cm.taskConfigManager.AddTask(uid, NewTaskConfig(pid, config))
	if err != nil {
		return err
	}

	processTask, err := config.New(values)
	if err != nil {
		_ = cm.taskConfigManager.RecoverTask(uid)
```
I am sorry to make function name confusing, taskConfigManager.RecoverTask actually recover the task config here , and taskConfigManager.AddTask add task config above.I will fix it.